codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoder.java

```diff
@@ -0,0 +1,402 @@
+            int calcMax = maxTlvSize + V2_MIN_LENGTH;
+            if (calcMax > V2_MAX_LENGTH) {
+                v2MaxHeaderSize = V2_MAX_LENGTH;
+            } else {
+                v2MaxHeaderSize = calcMax;
+            }
+        }
+    }
+
+    /**
+     * Returns the proxy protocol specification version in the buffer if the version is found.
+     * Returns -1 if no version was found in the buffer.
+     */
+    private static int findVersion(final ByteBuf buffer) {
+        final int n = buffer.readableBytes();
+        // per spec, the version number is found in the 13th byte
+        if (n < 13) {
+            return -1;
+        }
+
+        int idx = buffer.readerIndex();
+        return match(BINARY_PREFIX, buffer, idx) ? buffer.getByte(idx + BINARY_PREFIX_LENGTH) : 1;
+    }
+
+    /**
+     * Returns the index in the buffer of the end of header if found.
+     * Returns -1 if no end of header was found in the buffer.
+     */
+    private static int findEndOfHeader(final ByteBuf buffer) {
+        final int n = buffer.readableBytes();
+
+        // per spec, the 15th and 16th bytes contain the address length in bytes
+        if (n < 16) {
+            return -1;
+        }
+
+        int offset = buffer.readerIndex() + 14;
+
+        // the total header length will be a fixed 16 byte sequence + the dynamic address information block
+        int totalHeaderBytes = 16 + buffer.getUnsignedShort(offset);
+
+        // ensure we actually have the full header available
+        if (n >= totalHeaderBytes) {
+            return totalHeaderBytes;
+        } else {
+            return -1;
+        }
+    }
+
+    /**
+     * Returns the index in the buffer of the end of line found.
+     * Returns -1 if no end of line was found in the buffer.
+     */
+    private static int findEndOfLine(final ByteBuf buffer) {
+        final int n = buffer.writerIndex();
+        for (int i = buffer.readerIndex(); i < n; i++) {
+            final byte b = buffer.getByte(i);
+            if (b == '\r' && i < n - 1 && buffer.getByte(i + 1) == '\n') {
+                return i;  // \r\n
+            }
+        }
+        return -1;  // Not found.
+    }
+
+    @Override
+    public boolean isSingleDecode() {
+        // ByteToMessageDecoder uses this method to optionally break out of the decoding loop after each unit of work.
+        // Since we only ever want to decode a single header we always return true to save a bit of work here.
+        return true;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        super.channelRead(ctx, msg);
+        if (finished) {
+            ctx.pipeline().remove(this);
+        }
+    }
+
+    @Override
+    protected final void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        // determine the specification version
+        if (version == -1) {
+            if ((version = findVersion(in)) == -1) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1066)